### ISSUE TITLE
regression alerting: Long-term queryable records of alerts

### ISSUE THREAD

**Issue by alistaire47** [conbench/conbench#1249]
Distilling a conversation from last week with @jonkeane into a problem statement and proposed solution so we can have any necessary conversation to get on the same page for approach:

### Problem

Currently, if Conbench users (superusers? admins? devs?) want to know how many regression alerts were raised on runs (merge-commit is maybe most relevant here, though any reason is valid and maybe interesting) for a repo within some period of time (say 3 months), there is not a trivial way to obtain that information. 

Current options include 

1. Scraping the GitHub API and looking for and grepping comments on PRs.
    - Advantages: Directly corresponds to what developers are notified of
    - Disadvantages: Comments may get removed; text structure may change making maintaining a script solution for this approach a pain
2. Hitting the `/api/compare` endpoint for runs of interest and aggregating results
    - Advantages: Up-to-date information; not subject to inconsistencies of GitHub workflows
    - Disadvantages: Potentially a huge number of API calls and data processing; if rerun with any regularity will result in running the same calculations repeatedly

### Solution requirements

1. Is for alerts at a configurable threshold, parallel to all our alerting
2. Is easy to query. It should be trivial to group by conceptual benchmark, hardware, tags, etc. Schema should be consistent over time.
3. Will work for future alerts. Does not necessarily need to work for currently historical alerts.

### Proposed solution

A benchalerts step that will append new regressions to a file in an S3 (or similar) bucket (or configurable location). Structure of results is presumably close to the `/api/compare` structure or [benchalerts' internal data structures](https://github.com/conbench/conbench/blob/9f31f6b9f90a58ca084e9b8fda0147f69eca9145/benchalerts/benchalerts/conbench_dataclasses.py), but should be consistent even if those structures change.

### Discussion

1. Long-term, it would be better to make this data available via the API and UI. Are there ways to do so now without significant lifts?
2. If not, are there ways to structure this solution such that it will be more compatible with eventual server-side solutions?
3. Is S3 where we want this data? We have data in lots of places (Postgres, data warehouse, S3, GitHub, Slack, etc.). S3 is easy to set up. Is it where we want things over the medium-to-long term this solution is solving for?
4. What is the query latency requirement of this solution? How up-to-date does data need to be?

**Comment by austin3dickey** [comment]
>Long-term, it would be better to make this data available via the API and UI.

I agree; this is an important topic we've discussed in-person in the past (I need to make an issue). For a system like Conbench, it makes a lot of sense to centralize alerting on the server side. We have the `benchalerts` CI library today because we need to iterate quickly and respond to feedback from our various projects. But in the future, I think an ideal state looks like
- the server is responsible for making the HTTP API calls to various services to alert
- the database keeps track of past & current alerts
- this data is accessible via API
- all alerting parameters (thresholds, messages, locations) are configurable via a page in the webapp, and that changes the webapp experience as well (see my comment [here](https://github.com/conbench/conbench/issues/1252#issuecomment-1540147228))

This would ensure there's exactly one source of truth for managing alerts (and triaging them). [Arrow-benchmarks-ci](https://github.com/voltrondata-labs/arrow-benchmarks-ci) comes close to managing state in a database like this, but it's still a separate solution from the server.

When designing `benchalerts` features, we've ensured all the work we've done isn't a wasted effort; almost all the code in there now should still be useful at the time we make this switch.

---

With that said: I don't think this proposed solution will be useful toward that long-term goal:

>A benchalerts step that will append new regressions to a file in an S3

I do agree that we need to quickly solve the problem of "I'd like to be able to see at a glance how many times we've analyzed (and/or alerted on) runs." But I don't think adding a step like that is useful to the `benchalerts` mission.

As a reminder, `benchalerts` does already output all its raw, structured data from the `AlertPipeline`. That output structure is relatively stable (and CI systems using it should be pinned to a specific version of `benchalerts` anyway while we're working on the interface). So the only remaining step is to store that structured data somewhere long-term. My hot take: since every CI system is different and has access to different resources, we shouldn't handle that storage logic in `benchalerts`. Instead, the CI system should ETL that data itself.

It should be really simple to extract the JSON you need from that output and store it in S3 (or somewhere else), but I'd rather we put that code in the CI repo for now, since it's not necessary for Arrow and longer-term we want to do something completely different. Another benefit is that the CI job can be in full control of handling errors in that process and its corresponding complexity.

**Comment by jonkeane** [comment]
Thanks for all this summary + extensions on what we've chatted about + other thoughts in here. I'm excited to get something going where we can monitor this better.

> Currently, if Conbench users (superusers? admins? devs?) want to know how many regression alerts were raised on runs (merge-commit is maybe most relevant here, though any reason is valid and maybe interesting) for a repo within some period of time (say 3 months), there is not a trivial way to obtain that information.

IMO, we should develop something today that targets only conbench devs. That is where we currently don't have visibility and need to have some idea about how frequently we are pulling the various alarm bells we have. We will _eventually_ take lessons we learn from using this ourselves to turn it into something that might be user or superuser facing. But thinking about those folks + that design right now is premature. We've got to use it ourselves + find the rough edges + eliminate lots of the solution space before we plan out even what | where | how we ultimately will build as an end-user thing. And who knows, we might never build that thing! We might find the alerts are all totally just fine + expected and nothing is wrong. Or we might figure out our algos are funky with x, y, z edge case and we can refine that and fix it and everything is fine. Or maybe we do end up wanting to expose this to other users of conbench because it's helpful.


> Is easy to query. It should be trivial to group by conceptual benchmark, hardware, tags, etc. Schema should be consistent over time.

The schema being consistent is nice, but that is not a literal requirement. There are query tools that can deal with some kinds of schema evolution (and specifically the one we are most likely to experience: adding columns). Part of this project is about figuring out _what the real schema_ we need is. And to get to that, we need to use the data, see where it's lacking, and iterate. That's ok — and in fact will get us to something useful faster than trying to define a full possible schema today and not move forward until we have that.
 
 > Will work for future alerts. Does not necessarily need to work for currently historical alerts.

💯 Yes on this, just to double underline: though it would be _fantastic_ to have historical alerts, let's build the simplest thing we can looking forward. If we want later, we can add historical ones, but that's not a requirement (and frankly we would do ourselves a favor to even ignore it as a possible stretch feature here!)

> Long-term, it would be better to make this data available via the API and UI. Are there ways to do so now without significant lifts?
> If not, are there ways to structure this solution such that it will be more compatible with eventual server-side solutions?

I frankly think that trying to optimize for the possible future will make this task harder, longer, and more frustrating than it needs to be. IMO we should try and put these kinds of questions and possibilities out of our head. IF we have to rewrite this entire thing when we pull it into the server that is 💯 ok! We will get to a better solution faster if we build a thing that helps us today, figure out what the issues are and then build something entirely separate tomorrow for helping others (of course, if there's some code reuse, that's great, but truly, I believe if we ignore this and proceed iteratively even if we write the thing all over again twice that will be both quicker and lead to a better end product than trying to design something today that will eventually evolve into the end thing we think we want)

> Is S3 where we want this data? We have data in lots of places (Postgres, data warehouse, S3, GitHub, Slack, etc.). S3 is easy to set up. Is it where we want things over the medium-to-long term this solution is solving for?

S3 (or GCS) is the easiest, let's do that for now and re-assess if we have issues or we start thinking about exposing this beyond us.

> What is the query latency requirement of this solution? How up-to-date does data need to be?

Latency can be effectively infinite IMO. This is a human wants a result + analysis, not some system needs this and is blocked waiting. Up-to-date: as much as possible. Ideally as soon as a benchalert is posted, we know about it. But even if that were a day-delayed that's _probably fine_.


From Austins comments:

> Long-term, it would be better to make this data available via the API and UI.
> [and following]

If it helps us to defer this stuff, let's open a separate issue for the (far in the future from today) exposing any of this outside | moving alters to the server | etc.


On the CI versus benchalerts point:

I can see both sides to this, frankly. 

Though we have not been concentrating much effort on our apache/arrow instance, it would be good for us to be collecting data about how frequently that's alerting. For example: https://github.com/voltrondata-labs/arrow-benchmarks-ci/pull/109 where I removed the M1 Mac from alerts becausBue they are _incredibly noisy_ to figure this out, I had to search through arrow PRs and see that we were alerting "high regressions!" on something like > 75% of PRs merged (so understandably, the arrow maintainers have started ignoring even our worst-of-the-worst messaging since it's near constant). This didn't take "too long" for me, but it was all manual searching + clicking through conbench, but definitely longer than it needs to + I didn't record the data anywhere, it was mostly eyeballing + confirming suspicions. So, if this were in benchalerts (_and_ apache/arrow used benchalerts!) we could deploy our test implementation to more places more easily.

But, I also agree + suspect that doing this in the CI portion (for now?) at least will be quicker to getting us data (which IMO is the most important thing we can do right now: get to a place where we can incrementally understand what's happening).

If I had to pick one or be a tie breaker, I would probably say let's do CI for now — but I'm not strongly attached to that and could see if being useful for now / the medium term in benchalerts too
